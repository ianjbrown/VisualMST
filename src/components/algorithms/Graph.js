import { Random } from "random-js";
// create a graph class
class Graph {
    // constructor
    // initialize no of vertices at 0 and adjacencylist as an empty map
    constructor(noOfVertices) {
        this.noOfVertices = noOfVertices;
        this.AdjList = new Map();
    };
    
    setNoOfVertices(noOfVertices) {
        this.noOfVertices = noOfVertices;
    }

    // adds vertex v and map to a null array
    addVertex(v) {
        this.AdjList.set(v, []);
        //this.state.noOfVertices += 1;
    }
    
    // add edge to each others adjacency lists since undirected graph
    addEdge(u, v, weight) {
        let uAdjList = this.AdjList.get(u);
        let vAdjList = this.AdjList.get(v);

        // no duplicate edges
        for (var i in uAdjList) if (uAdjList[i].node === v) return false;
        
        // no self-connecting edges
        if (u === v) return false;

        // undirected graph so push onto both
        uAdjList.push({
            node: v,
            weight: weight
        })
        vAdjList.push({
            node: u,
            weight: weight
        })
        return true;
    }
    
    /* this is really just a depth first search that returns
    true if we can visit all vertices (graph is connected)*/
    isConnected(startingNode) {
        var visited = {};
        this.isConnectedUtil(startingNode, visited);
        if (Object.keys(visited).length < this.noOfVertices) {
            return false;
        } else {
            return true;
        }
    }

    // recursive function for above
    isConnectedUtil(vert, visited) {
        visited[vert] = true;

        var get_neighbours = this.AdjList.get(vert);
        for (var i in get_neighbours) {
            var get_elem = get_neighbours[i].node;
            if (!visited[get_elem]) {
                this.isConnectedUtil(get_elem, visited);
            }
        }
    }

    toString() {
        // get all vertices
        var get_keys = this.AdjList.keys();
        
        var outerConc = [];
        // iterate over vertices
        for (var i of get_keys) {

            // get corresponsing adjacency list
            var get_values = this.AdjList.get(i);
            var conc = "";

            // iterate over adjacency list
            // concatenate the values into a string
            for (var j of get_values) {
                conc += "{ node = " + j.node + ", weight = " + j.weight + "}";
            }

            // print the vertex and its adjacency list
            outerConc.push(i + " -> " + conc);
        }
        console.log(this.AdjList);
        return outerConc;
        
    }

    generateGraph() {
        const random = new Random();
        let edges = 0;

        //Add the vertices
        for (var i = 0; i < this.noOfVertices; i++) {
            this.addVertex(i);
        }

        //Add edges until we have a connected graph
        while (!this.isConnected(0)) {
            //const edges = random.integer(1,)
            let added = this.addEdge(random.integer(0,this.noOfVertices - 1), random.integer(0,this.noOfVertices - 1), random.integer(1, 10));
            if (added) edges++;
        }   

        //add a random amount of more edges to allow for more than one MST
        let moreEdges = random.integer(edges, (this.noOfVertices * ((this.noOfVertices - 1) / 2)));
        while (edges < moreEdges) {
            let wasAdded = this.addEdge(random.integer(0,this.noOfVertices - 1), random.integer(0,this.noOfVertices - 1), random.integer(1, 10));
            if (wasAdded) edges++;
        }
    }
    
}
export default Graph;