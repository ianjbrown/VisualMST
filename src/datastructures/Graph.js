import { Random } from "random-js";
import PriorityQueue from "./PriorityQueue.js";
import UnionFind from "./UnionFind.js";
import AlgData from "./AlgData.js";
// create a graph class
class Graph {
  // constructor
  // initialize no of vertices at 0 and adjacencylist as an empty map
  constructor(noOfVertices) {
    this.noOfVertices = noOfVertices;
    this.AdjList = new Map();
  }

  setNoOfVertices(noOfVertices) {
    this.noOfVertices = noOfVertices;
  }

  // adds vertex v and map to a null array
  addVertex(v) {
    this.AdjList.set(v, []);
    //this.state.noOfVertices += 1;
  }

  getEdgeCount() {
    let count = 0;
    for (const [key, value] of this.AdjList.entries()) {
      count += value.length;
    }
    return count / 2;
  }

  // add edge to each others adjacency lists since undirected graph
  addEdge(u, v, weight) {
    let uAdjList = this.AdjList.get(u);
    let vAdjList = this.AdjList.get(v);

    if (uAdjList === undefined || vAdjList === undefined) return false;
    // no duplicate edges
    for (var i in uAdjList) if (uAdjList[i].node === v) return false;

    // no self-connecting edges
    if (u === v) return false;
    // undirected graph so push onto both
    uAdjList.push({
      node: v,
      weight: weight,
    });
    vAdjList.push({
      node: u,
      weight: weight,
    });
    return true;
  }

  /* this is really just a depth first search that returns
    true if we can visit all vertices (graph is connected)*/
  isConnected(startingNode) {
    var visited = {};
    this.isConnectedUtil(startingNode, visited);
    if (Object.keys(visited).length < this.noOfVertices) {
      return false;
    } else {
      return true;
    }
  }

  // recursive function for above
  isConnectedUtil(vert, visited) {
    visited[vert] = true;

    var get_neighbours = this.AdjList.get(vert);
    for (var i in get_neighbours) {
      var get_elem = get_neighbours[i].node;
      if (!visited[get_elem]) {
        this.isConnectedUtil(get_elem, visited);
      }
    }
  }

  toString() {
    // get all vertices
    var get_keys = this.AdjList.keys();

    var outerConc = [];
    var edgeCount = 0;
    // iterate over vertices
    for (var i of get_keys) {
      // get corresponsing adjacency list
      var get_values = this.AdjList.get(i);
      var conc = "";

      // iterate over adjacency list
      // concatenate the values into a string
      for (var j of get_values) {
        conc += "{ node = " + j.node + ", weight = " + j.weight + "}";
        edgeCount++;
      }

      // print the vertex and its adjacency list
      outerConc.push(i + " -> " + conc);
    }
    outerConc.push(
      "This Graph has " +
        this.noOfVertices +
        " vertices and " +
        edgeCount / 2 +
        " edges"
    );
    if (edgeCount / 2 === this.noOfVertices - 1) {
      outerConc.push(
        "This Graph is an MST because it has n-1 (" + edgeCount / 2 + ") edges"
      );
    }
    return outerConc;
  }

  generateGraph(lines) {
    const random = new Random();
    for (var i = 0; i < this.noOfVertices; i++) {
      this.addVertex(i);
    }

    for (i = 0; i < lines.length; i++) {
      const edge = lines[i].split(",");
      this.addEdge(parseInt(edge[0]), parseInt(edge[1]), random.integer(1, 30));
    }
  }

  // Generate random graph, app now generates predetermined graphs from backend instead.
  generateRandomGraph() {
    const random = new Random();
    let edges = 0;

    this.AdjList.clear();

    //Add the vertices
    for (var i = 0; i < this.noOfVertices; i++) {
      this.addVertex(i);
    }

    //Add edges until we have a connected graph
    while (!this.isConnected(0)) {
      //const edges = random.integer(1,)
      let added = this.addEdge(
        random.integer(0, this.noOfVertices - 1),
        random.integer(0, this.noOfVertices - 1),
        random.integer(1, 30)
      );
      if (added) edges++;
    }

    // add a random amount of more edges to allow for more than one MST
    let moreEdges = random.integer(
      edges + 1,
      0.25 * this.noOfVertices * (this.noOfVertices - 1)
    );
    while (edges < moreEdges) {
      let wasAdded = this.addEdge(
        random.integer(0, this.noOfVertices - 1),
        random.integer(0, this.noOfVertices - 1),
        random.integer(1, 30)
      );
      if (wasAdded) edges++;
    }
  }

  importGraph(lines) {
    //add the vertices
    for (var i = 0; i < this.noOfVertices; i++) {
      this.addVertex(i);
    }

    //add the edges
    for (i = 0; i < lines.length; i++) {
      const edge = lines[i].split(",");
      this.addEdge(parseInt(edge[0]), parseInt(edge[1]), parseInt(edge[2]));
    }
  }

  kruskal() {
    // Create new graph for MST and priority queue for edges sorted by weight
    // create union-find/disjoint-sets data structure to help check for introduction of cycles
    var edgeCompSequence = [];
    var edgesQueues = [];
    var minWeight = 0;
    var edgeSequence = [];
    const MST = new Graph(this.noOfVertices);
    var edgesQueue = new PriorityQueue();
    let uf = new UnionFind(this.AdjList);
    var ufs = [uf];

    // Add vertices to MST
    for (var i = 0; i < this.noOfVertices; i++) {
      MST.addVertex(i);
    }
    MST.setNoOfVertices(MST.AdjList.size);
    // Add edges to priority queue
    for (const element of this.AdjList) {
      for (const adjacent of element[1]) {
        edgesQueue.enqueue([element[0], adjacent.node], adjacent.weight);
      }
    }
    edgesQueues.push(Object.assign([], edgesQueue.items));

    while (!edgesQueue.isEmpty()) {
      if (MST.getEdgeCount() === this.noOfVertices - 1) {
        // console.log("Terminate early");
        break;
      }
      const nextEdge = edgesQueue.dequeue();
      const vertices = nextEdge.elem;
      const weight = nextEdge.prio;
      ufs.push(uf);

      //if adding edge would not create cycle, add edge to MST
      if (!uf.connected(vertices[0], vertices[1])) {
        minWeight += weight;
        MST.addEdge(vertices[0], vertices[1], weight);
        uf.union(vertices[0], vertices[1]);
        edgeSequence.push([vertices[0], vertices[1], weight]);
      }
      edgesQueues.push(Object.assign([], edgesQueue.items));
      edgeCompSequence.push(nextEdge);
    }
    return new AlgData(MST, edgeSequence, edgesQueues, minWeight, edgeCompSequence);
    // return [MST, edgeSequence, edgesQueues, minWeight, edgeCompSequence];
  }

  prim(startingVertex) {
    var edgesQueues = [];
    var visiteds = [];
    var minWeight = 0;
    /* create data structures... new graph for MST, 
    queue for edges sorted by weight, and array of visited vertices */
    var edgeSequence = [];
    var edgeCompSequence = [];
    const MST = new Graph(this.noOfVertices);
    var edgesQueue = new PriorityQueue();
    var visited = [];

    //begin with first node (0)
    visited.push(startingVertex);
    MST.addVertex(startingVertex);

    // add edges of the starting node to queue
    for (const adjacent of this.AdjList.get(startingVertex)) {
      edgesQueue.enqueue([startingVertex, adjacent.node], adjacent.weight);
    }
    visiteds.push(Object.assign([], visited));
    edgesQueues.push(Object.assign([], edgesQueue.items));

    // while we have not visited all vertices
    while (visited.length < this.noOfVertices) {
      // pick edge with the lowest weight
      let nextEdge = edgesQueue.dequeue();

      // the connected vertex
      let nextNode = nextEdge.elem[1];

      // if not visited yet, add to MST
      if (!visited.includes(nextNode)) {
        MST.addVertex(nextNode);
        MST.addEdge(nextEdge.elem[0], nextNode, nextEdge.prio);
        minWeight += nextEdge.prio;
        edgeSequence.push([nextEdge.elem[0], nextNode, nextEdge.prio]);

        // add edges of new node to queue
        for (const adjacent of this.AdjList.get(nextNode)) {
          if (adjacent.node === nextEdge.elem[0]) continue;
          edgesQueue.enqueue([nextNode, adjacent.node], adjacent.weight);
        }

        // mark as visited
        visited.push(nextNode);
      }
      visiteds.push(Object.assign([], visited));
      edgesQueues.push(Object.assign([], edgesQueue.items));
      edgeCompSequence.push(nextEdge);

      MST.setNoOfVertices(MST.AdjList.size);
    }
    return new AlgData(
      MST,
      edgeSequence,
      edgesQueues,
      minWeight,
      edgeCompSequence,
      visiteds,
    );
    // return [
    //   MST,
    //   edgeSequence,
    //   edgesQueues,
    //   visiteds,
    //   minWeight,
    //   edgeCompSequence,
    // ];
  }

  // this helper function allows us to sort adjlists by numerical node value
  compareAdjListEntry(x, y) {
    if (x.node < y.node) {
      return -1;
    }
    if (x.node > y.node) {
      return 1;
    }
    return 0;
  }

  compareObjects(x, y) {
    for (var property in x) {
      if (x[property] !== y[property]) {
        return false;
      }
    }
    return true;
  }

  compareAdjLists(kruskalList, primList) {
    var testVal;
    if (kruskalList.size !== primList.size) {
      return false;
    }
    for (var [key, val] of kruskalList) {
      testVal = primList.get(key);
      if (!primList.has(key) || testVal.length !== val.length) {
        return false;
      }
      val.sort(this.compareAdjListEntry);
      testVal.sort(this.compareAdjListEntry);

      for (var i = 0; i < val.length; i++) {
        if (!this.compareObjects(val[i], testVal[i])) {
          return false;
        }
      }
    }
    return true;
  }
}
export default Graph;
