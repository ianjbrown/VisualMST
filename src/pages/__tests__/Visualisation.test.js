import VisualisationPage from "../Visualisation";
import Graph from "../../datastructures/Graph";
import { BrowserRouter as Router } from "react-router-dom";
import { render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";

afterEach(() => {
  cleanup();
});

test("Kruskal Visualisation page matches snapshot", () => {
  const g = new Graph(5);
  g.generateGraph(["0,1", "0,4", "1,2", "0,2", "2,3", "2,4", "3,4"]);
  const tree = renderer
    .create(
      <Router>
        <VisualisationPage alg="kruskal" graph={g} MSTGraph={g.kruskal()} />
      </Router>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

//   test("Prim Visualisation page matches snapshot", () => {
//     const g = new Graph(5);
//     g.generateGraph(["0,1","0,4","1,2","0,2","2,3","2,4","3,4"]);
//     const tree = renderer
//       .create(
//         <Router>
//           <VisualisationPage alg="prim" graph={g} MSTGraph={g.prim("0")} />
//         </Router>
//       )
//       .toJSON();
//     expect(tree).toMatchSnapshot();
//   });

// test("Prim's Visualisation loads", () => {
//   const g = new Graph(5);
//   g.generateGraph(["0,1", "0,4", "1,2", "0,2", "2,3", "2,4", "3,4"]);
//   render(
//     <Router>
//       <VisualisationPage alg="prim" graph={g} MSTGraph={g.prim("0")} />
//     </Router>
//   );

// });
