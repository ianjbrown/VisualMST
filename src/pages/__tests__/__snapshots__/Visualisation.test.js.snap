// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kruskal Visualisation page matches snapshot 1`] = `
Array [
  <h2
    className="pt-1"
  >
    Kruskal's Algorithm
  </h2>,
  <div
    className="closeButton"
  >
    <a
      className="btn btn-close"
      href="/"
      onClick={[Function]}
    />
  </div>,
  <div
    className="row"
  >
    <div
      className="algorithmStateMessage"
    >
      Edges are sorted by their weight in non-decreasing order
    </div>
  </div>,
  <div
    className="canvasDiv"
    style={
      Object {
        "left": "10px",
        "right": "500px",
      }
    }
  >
    <canvas
      className="canvasStyle"
    />
    <canvas
      className="canvasStyle"
    />
  </div>,
  <div
    className="infoPanelTab"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ArrowForwardIosOutlinedIcon"
      focusable="false"
      onClick={[Function]}
      viewBox="0 0 24 24"
    >
      <path
        d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
      />
    </svg>
  </div>,
  <div
    className="infoPanel"
    style={
      Object {
        "marginLeft": 0,
        "opacity": 1,
      }
    }
  >
    <h1>
      Pseudocode
    </h1>
    <ul
      className="list-group list-group-flush"
      data-testid="kruskal-list-group"
      onKeyDown={[Function]}
    >
      <li
        className="list-group-item list-group-item-warning"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
        Sort edges in non-decreasing order;
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
        While edges remain in queue
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
          e = edge with smallest weight; 
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
          If adding e to MST doesn't create cycle 
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
           Add e to MST;  
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
          Else 
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
           Reject e from MST;  
      </li>
      <li
        className="list-group-item list-group-item-dark"
        data-rr-ui-event-key={null}
        onClick={[Function]}
      >
        We have found a minimum spanning tree with n-1 edges
      </li>
    </ul>
    <h1>
      Edges Queue
    </h1>
    <div
      className="edgesQueueList"
    >
      <ol>
        <li>
          ((
          0
          ,
          1
          ). 
          5
          )
        </li>
        <li>
          ((
          0
          ,
          2
          ). 
          12
          )
        </li>
        <li>
          ((
          2
          ,
          3
          ). 
          15
          )
        </li>
        <li>
          ((
          3
          ,
          4
          ). 
          18
          )
        </li>
        <li>
          ((
          0
          ,
          4
          ). 
          19
          )
        </li>
        <li>
          ((
          1
          ,
          2
          ). 
          20
          )
        </li>
        <li>
          ((
          2
          ,
          4
          ). 
          25
          )
        </li>
      </ol>
    </div>
  </div>,
  <div
    className="fixedRowBottom row"
  >
    <div
      className="bar col-3"
      data-testid="slider-col"
    >
      <div
        className="barText"
        data-testid="slow"
      >
        Slow
      </div>
      <span
        className="MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-root slider css-187mznn-MuiSlider-root"
        data-testid="slider"
        onMouseDown={[Function]}
      >
        <span
          className="MuiSlider-rail css-14pt78w-MuiSlider-rail"
        />
        <span
          className="MuiSlider-track css-1gv0vcd-MuiSlider-track"
          style={
            Object {
              "left": "0%",
              "width": "47.36842105263158%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={0}
          style={
            Object {
              "left": "0%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={1}
          style={
            Object {
              "left": "5.2631578947368425%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={2}
          style={
            Object {
              "left": "10.526315789473685%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={3}
          style={
            Object {
              "left": "15.789473684210526%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={4}
          style={
            Object {
              "left": "21.05263157894737%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={5}
          style={
            Object {
              "left": "26.31578947368421%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={6}
          style={
            Object {
              "left": "31.57894736842105%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={7}
          style={
            Object {
              "left": "36.8421052631579%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={8}
          style={
            Object {
              "left": "42.10526315789474%",
            }
          }
        />
        <span
          className="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
          data-index={9}
          style={
            Object {
              "left": "47.36842105263158%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={10}
          style={
            Object {
              "left": "52.63157894736842%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={11}
          style={
            Object {
              "left": "57.89473684210526%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={12}
          style={
            Object {
              "left": "63.1578947368421%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={13}
          style={
            Object {
              "left": "68.42105263157895%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={14}
          style={
            Object {
              "left": "73.6842105263158%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={15}
          style={
            Object {
              "left": "78.94736842105263%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={16}
          style={
            Object {
              "left": "84.21052631578948%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={17}
          style={
            Object {
              "left": "89.47368421052632%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={18}
          style={
            Object {
              "left": "94.73684210526316%",
            }
          }
        />
        <span
          className="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
          data-index={19}
          style={
            Object {
              "left": "100%",
            }
          }
        />
        <span
          className="MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium MuiSlider-thumb css-eg0mwd-MuiSlider-thumb"
          data-index={0}
          onMouseLeave={[Function]}
          onMouseOver={[Function]}
          style={
            Object {
              "left": "47.36842105263158%",
              "pointerEvents": undefined,
            }
          }
        >
          <input
            aria-label="Speed"
            aria-orientation="horizontal"
            aria-valuemax={2000}
            aria-valuemin={100}
            aria-valuenow={1000}
            data-index={0}
            disabled={false}
            max={2000}
            min={100}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            step={100}
            style={
              Object {
                "border": 0,
                "clip": "rect(0 0 0 0)",
                "direction": "ltr",
                "height": "100%",
                "margin": -1,
                "overflow": "hidden",
                "padding": 0,
                "position": "absolute",
                "whiteSpace": "nowrap",
                "width": "100%",
              }
            }
            type="range"
            value={1000}
          />
          <span
            aria-hidden={true}
            className="MuiSlider-valueLabel css-1kz0hui-MuiSlider-valueLabel"
          >
            <span
              className="MuiSlider-valueLabelCircle"
            >
              <span
                className="MuiSlider-valueLabelLabel"
              >
                <div>
                  Speed: 
                  5
                </div>
              </span>
            </span>
          </span>
        </span>
      </span>
      <div
        className="barText"
        data-testid="fast"
      >
        Fast
      </div>
    </div>
    <div
      className="bar col-6"
      data-testid="playback-col"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
        data-testid="FirstPageIcon"
        focusable="false"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M18.41 16.59 13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
        data-testid="ArrowBackIosNewOutlinedIcon"
        focusable="false"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M17.77 3.77 16 2 6 12l10 10 1.77-1.77L9.54 12z"
        />
      </svg>
      <div
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
          data-testid="PlayArrowIcon"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M8 5v14l11-7z"
          />
        </svg>
      </div>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
        data-testid="ArrowForwardIosOutlinedIcon"
        focusable="false"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M6.23 20.23 8 22l10-10L8 2 6.23 3.77 14.46 12z"
        />
      </svg>
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
        data-testid="LastPageIcon"
        focusable="false"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
        viewBox="0 0 24 24"
      >
        <path
          d="M5.59 7.41 10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
        />
      </svg>
    </div>
    <div
      className="bar col-3"
      data-testid="toggle-col"
    >
      <div
        className="barToggle"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1tvxjov-MuiSvgIcon-root"
          data-testid="ToggleOffOutlinedIcon"
          focusable="false"
          onClick={[Function]}
          viewBox="0 0 24 24"
        >
          <path
            d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
          />
        </svg>
      </div>
      <div
        className="barText"
        data-testid="hide-text"
      >
        Hide Original Graph
      </div>
    </div>
  </div>,
]
`;
